File: reference/python/crovia_seal/canonical.py
"""
CSC-1 — Crovia Seal Canonicalization v1.

Produces a fully deterministic UTF-8 byte representation of a JSON value.
CSC-1 is a strict subset of RFC 8785 (JCS): it adopts JCS's ordering and
string-escape rules, and FORBIDS floating-point numbers in signed payloads.
This eliminates the dependency on ECMA-262 7.1.12.1 number serialization,
which is the main source of cross-language portability risk in JCS.

Forbidding floats is not a practical restriction: the Seal specification
requires that all parameter values carried inside the signed payload be
encoded as strings (see SPEC.md Section 4.6).

Any input that would yield a non-canonical form raises a subclass of
CanonicalizationError. There is no "best effort" mode: CSC-1 is fail-closed
by design.
"""
from __future__ import annotations

from typing import Any, List

from crovia_seal.constants import JS_SAFE_INT_MAX, JS_SAFE_INT_MIN
from crovia_seal.errors import (
    DuplicateKey,
    NonCanonicalNumber,
    NonStringKey,
    UnsupportedType,
)


# --- String serialization ---------------------------------------------------

# Table of mandatory escape sequences per RFC 8259. All other characters
# outside the U+0000..U+001F range are emitted literally.
_ESCAPE_MAP = {
    0x22: '\\"',   # "
    0x5C: "\\\\",  # \
    0x08: "\\b",
    0x0C: "\\f",
    0x0A: "\\n",
    0x0D: "\\r",
    0x09: "\\t",
}


def _serialize_string(s: str) -> str:
    """Serialize a Python str to a canonical JSON string literal.

    Follows RFC 8785 Section 3.2.2.2: the seven short escapes (",\,b,f,n,r,t)
    are used for the corresponding code points; any other control character
    in U+0000..U+001F uses the \\u00XX form; all code points >= U+0020 are
    emitted literally (including non-ASCII, which will be UTF-8-encoded by
    the caller). Surrogate pairs are preserved as-is.
    """
    out: List[str] = ['"']
    for ch in s:
        cp = ord(ch)
        if cp < 0x20:
            esc = _ESCAPE_MAP.get(cp)
            if esc is not None:
                out.append(esc)
            else:
                out.append(f"\\u{cp:04x}")
        elif cp == 0x22 or cp == 0x5C:
            # These two are escaped even though >= 0x20
            out.append(_ESCAPE_MAP[cp])
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


# --- Number serialization (CSC-1: integers only) ----------------------------

def _serialize_number(n: int) -> str:
    """Serialize a Python int to its shortest decimal form.

    CSC-1 forbids floats; callers MUST pass int. We also enforce the
    JavaScript-safe integer range for maximum cross-language portability,
    matching RFC 8785 practice for whole numbers.

    Python `bool` is a subclass of int; we reject it here to avoid
    `True` sneaking through as `1`. Booleans must take the true/false path.
    """
    if isinstance(n, bool):
        # bool MUST be handled by the dispatcher, not here. Raising here
        # guards against mis-routing.
        raise UnsupportedType("bool must be serialized as true/false, not as int")
    if not isinstance(n, int):
        raise NonCanonicalNumber(
            f"CSC-1 permits only int numbers; got {type(n).__name__}"
        )
    if n < JS_SAFE_INT_MIN or n > JS_SAFE_INT_MAX:
        raise NonCanonicalNumber(
            f"integer {n} is outside the JS-safe range "
            f"[{JS_SAFE_INT_MIN}, {JS_SAFE_INT_MAX}]; "
            "encode large integers as strings"
        )
    # Python's str(int) already produces the shortest form with no leading
    # zeros and an explicit leading minus sign for negatives. For 0 it
    # produces "0" (not "-0"), which matches CSC-1 requirements.
    return str(n)


# --- Array / object serialization -------------------------------------------

def _serialize_array(arr: list) -> str:
    parts = [_serialize(v) for v in arr]
    return "[" + ",".join(parts) + "]"


def _serialize_object(obj: dict) -> str:
    # Reject non-string keys. Python dicts allow e.g. int keys, which JSON
    # does not. Silently stringifying them would mask a bug.
    for k in obj.keys():
        if not isinstance(k, str):
            raise NonStringKey(
                f"object key must be str, got {type(k).__name__}: {k!r}"
            )

    # RFC 8785 Section 3.2.3: keys are sorted ascending by UTF-16 code unit.
    # Python strings are Unicode scalar values; `sorted(..., key=lambda k: k.encode('utf-16-be'))`
    # gives the exact UTF-16 code-unit order. For BMP-only keys (the common
    # case), this is equivalent to `sorted(keys)`, but we use the explicit
    # encoding to be correct on supplementary-plane code points.
    def _utf16_key(k: str) -> bytes:
        return k.encode("utf-16-be")

    # Detect duplicate keys as a defense against dict constructions that
    # could merge silently. (A Python dict literal with duplicate keys
    # keeps the last one; we want to flag this.)
    seen = set()
    ordered_keys = []
    for k in obj.keys():
        if k in seen:
            raise DuplicateKey(f"duplicate object key: {k!r}")
        seen.add(k)
        ordered_keys.append(k)

    ordered_keys.sort(key=_utf16_key)

    parts = []
    for k in ordered_keys:
        parts.append(_serialize_string(k) + ":" + _serialize(obj[k]))
    return "{" + ",".join(parts) + "}"


# --- Dispatcher -------------------------------------------------------------

def _serialize(value: Any) -> str:
    """Serialize a single JSON value per CSC-1 rules."""
    # Order matters: bool before int (bool is subclass of int in Python).
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, str):
        return _serialize_string(value)
    if isinstance(value, int):
        return _serialize_number(value)
    if isinstance(value, float):
        raise NonCanonicalNumber(
            "CSC-1 forbids float in signed payloads; "
            "encode numeric parameters as strings (e.g. temperature=\"0.7\")"
        )
    if isinstance(value, list) or isinstance(value, tuple):
        return _serialize_array(list(value))
    if isinstance(value, dict):
        return _serialize_object(value)
    raise UnsupportedType(
        f"CSC-1 cannot serialize value of type {type(value).__name__}"
    )


# --- Public API -------------------------------------------------------------

def canonicalize(value: Any) -> bytes:
    """Produce the CSC-1 UTF-8 byte sequence for the given JSON-like value.

    Accepts nested structures of: None, bool, int (within JS safe range),
    str, list/tuple, dict (with str keys only). Any other type — including
    float — raises CanonicalizationError.

    The output is deterministic: two Python objects that are logically equal
    as JSON will canonicalize to the same bytes. This is the property that
    makes signing meaningful.
    """
    text = _serialize(value)
    return text.encode("utf-8")

File: reference/typescript/src/canonical.ts
/**
 * CSC-1 - Crovia Seal Canonicalization v1.
 *
 * TypeScript port of `crovia_seal/canonical.py`. MUST produce byte-identical
 * output to the Python reference for all valid inputs. Divergence is a bug.
 *
 * Key correctness notes for JavaScript:
 *
 * - JS numbers are IEEE-754 doubles. CSC-1 forbids floats in signed
 *   payloads, so we accept ONLY `number` values satisfying
 *   `Number.isInteger(n) && Number.isSafeInteger(n)`. Anything else
 *   (fractional, NaN, +/-Infinity, out-of-range) is rejected.
 *
 * - BigInt: accepted IF within the JS-safe integer range (for cross-language
 *   compatibility with Python's int). Outside the range: rejected.
 *
 * - String iteration: JS `for..of` on a string iterates by Unicode scalar
 *   value (code point), which is what CSC-1 wants. Surrogate pairs in the
 *   source are presented as a single supplementary-plane code point.
 *
 * - Object key sorting: JS's default string comparison (`a < b`) operates
 *   on UTF-16 code units, which matches Python's `str.encode("utf-16-be")`
 *   byte comparison. For BMP-only keys this equals lexicographic sort; for
 *   supplementary-plane keys both sides agree on surrogate-pair ordering.
 */
import {
  DuplicateKey,
  NonCanonicalNumber,
  NonStringKey,
  UnsupportedType,
} from './errors.js';
import { JS_SAFE_INT_MAX, JS_SAFE_INT_MIN } from './constants.js';

/** JSON-like values we accept. */
export type JsonValue =
  | null
  | boolean
  | number
  | bigint
  | string
  | readonly JsonValue[]
  | { readonly [k: string]: JsonValue };


// --- String serialization --------------------------------------------------

const SHORT_ESCAPES: Record<number, string> = {
  0x08: '\\b',
  0x0c: '\\f',
  0x0a: '\\n',
  0x0d: '\\r',
  0x09: '\\t',
};

function _serializeString(s: string): string {
  let out = '"';
  // for..of iterates by Unicode code point, which is what we want.
  // This correctly handles supplementary-plane code points (emoji etc).
  for (const ch of s) {
    const cp = ch.codePointAt(0)!;
    if (cp < 0x20) {
      const esc = SHORT_ESCAPES[cp];
      if (esc !== undefined) {
        out += esc;
      } else {
        // \u00XX form, lowercase hex to match Python.
        out += '\\u' + cp.toString(16).padStart(4, '0');
      }
    } else if (cp === 0x22) {
      out += '\\"';
    } else if (cp === 0x5c) {
      out += '\\\\';
    } else {
      out += ch;
    }
  }
  out += '"';
  return out;
}


// --- Number serialization (CSC-1: integers only) ---------------------------

function _serializeNumber(n: number | bigint): string {
  if (typeof n === 'bigint') {
    if (n < BigInt(JS_SAFE_INT_MIN) || n > BigInt(JS_SAFE_INT_MAX)) {
      throw new NonCanonicalNumber(
        `bigint ${n} is outside the JS-safe range ` +
        `[${JS_SAFE_INT_MIN}, ${JS_SAFE_INT_MAX}]; ` +
        'encode large integers as strings',
      );
    }
    return n.toString(10);
  }

  if (!Number.isFinite(n)) {
    throw new NonCanonicalNumber(
      'CSC-1 forbids NaN and +/-Infinity',
    );
  }
  if (!Number.isInteger(n)) {
    throw new NonCanonicalNumber(
      'CSC-1 forbids float in signed payloads; ' +
      'encode numeric parameters as strings (e.g. temperature="0.7")',
    );
  }
  // Treat -0 as 0 (Python: str(0) == "0").
  if (Object.is(n, -0)) {
    return '0';
  }
  if (!Number.isSafeInteger(n)) {
    throw new NonCanonicalNumber(
      `integer ${n} is outside the JS-safe range ` +
      `[${JS_SAFE_INT_MIN}, ${JS_SAFE_INT_MAX}]; ` +
      'encode large integers as strings',
    );
  }
  // Number.prototype.toString(10) for a safe integer produces the shortest
  // decimal representation with no leading zeros and an explicit leading
  // minus sign for negatives. For 0 it produces "0".
  return n.toString(10);
}


// --- Array serialization ---------------------------------------------------

function _serializeArray(arr: readonly JsonValue[]): string {
  let out = '[';
  let first = true;
  for (const v of arr) {
    if (!first) out += ',';
    first = false;
    out += _serialize(v);
  }
  out += ']';
  return out;
}


// --- Object serialization --------------------------------------------------

/**
 * A "plain object" is one whose prototype is `Object.prototype` or `null`.
 * This deliberately excludes Uint8Array, Set, Map, Date, class instances,
 * and any other exotic object. Passing such objects to canonicalize() is
 * a type error, not a silent empty-object serialization.
 *
 * SECURITY: without this check, a Uint8Array serializes as
 * `{"0":n0,"1":n1,...}` and a Set serializes as `{}`. Both would produce
 * valid signatures over data the caller never intended to sign.
 */
function _isPlainObject(v: unknown): boolean {
  if (v === null || typeof v !== 'object') return false;
  const proto = Object.getPrototypeOf(v);
  return proto === Object.prototype || proto === null;
}

function _serializeObject(obj: { readonly [k: string]: JsonValue }): string {
  // Gather keys with JS semantics. `Object.keys` returns the own enumerable
  // string keys in definition order (ignoring Symbol keys).
  const keys = Object.keys(obj);

  // Duplicate-key detection: JS object literals cannot have duplicate keys
  // in the source (last wins). But if someone constructs a dict-like and
  // hands us an object with a collision that somehow survives, we still
  // want to flag it. In practice, once keys is an array of strings obtained
  // from Object.keys, duplicates cannot appear. We check defensively.
  const seen = new Set<string>();
  for (const k of keys) {
    if (typeof k !== 'string') {
      throw new NonStringKey(`object key must be string, got ${typeof k}`);
    }
    if (seen.has(k)) {
      throw new DuplicateKey(`duplicate object key: ${JSON.stringify(k)}`);
    }
    seen.add(k);
  }

  // Sort keys by UTF-16 code-unit sequence. JS string `<` already does this
  // natively (compares by UTF-16 code units). We use a stable sort.
  keys.sort();

  let out = '{';
  let first = true;
  for (const k of keys) {
    if (!first) out += ',';
    first = false;
    out += _serializeString(k) + ':' + _serialize(obj[k]!);
  }
  out += '}';
  return out;
}


// --- Dispatcher ------------------------------------------------------------

function _serialize(v: JsonValue): string {
  if (v === null) return 'null';
  if (v === true) return 'true';
  if (v === false) return 'false';
  const t = typeof v;
  if (t === 'string') return _serializeString(v as string);
  if (t === 'number' || t === 'bigint') return _serializeNumber(v as number | bigint);
  if (Array.isArray(v)) return _serializeArray(v);
  // `typeof` alone is insufficient: Uint8Array, Set, Map, Date etc. are all
  // `"object"` but must NOT be silently coerced. Require a plain object.
  if (t === 'object') {
    if (!_isPlainObject(v)) {
      const ctor = (v as { constructor?: { name?: string } }).constructor?.name ?? 'unknown';
      throw new UnsupportedType(
        `CSC-1 accepts only plain objects; got ${ctor}. ` +
        'Convert exotic values (Uint8Array, Set, Map, Date, ...) to a supported ' +
        'representation (array, string, integer) before canonicalizing.',
      );
    }
    return _serializeObject(v as { [k: string]: JsonValue });
  }
  throw new UnsupportedType(`CSC-1 cannot serialize value of type ${t}`);
}


// --- Public API ------------------------------------------------------------

/**
 * Produce the CSC-1 UTF-8 byte sequence for the given JSON-like value.
 *
 * Accepts nested structures of: null, boolean, number (integer only, JS-safe
 * range), bigint (JS-safe range), string, array, plain object (string keys).
 * Any other type - including float - throws a CanonicalizationError subclass.
 *
 * The output is deterministic: two values that are logically equal as JSON
 * will canonicalize to the same bytes. This is what makes signing meaningful.
 */
export function canonicalize(value: JsonValue): Uint8Array {
  const text = _serialize(value);
  return new TextEncoder().encode(text);
}
